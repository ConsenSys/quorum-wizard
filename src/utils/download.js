import { basename, join } from 'path'
import { createFolder, cwd, exists } from './fileUtils'
import axios from 'axios'
import { createGunzip } from 'zlib'
import { extract } from 'tar-fs'
import { createWriteStream } from 'fs'
import { BINARIES, getPlatformSpecificUrl } from './binaryHelper'

export async function downloadIfMissing (name, version) {
  if (BINARIES[name] === undefined || BINARIES[name][version] === undefined) {
    throw new Error(`Could not find binary info entry for ${name} ${version}`)
  }
  const binaryInfo = BINARIES[name][version]
  const binDir = join(cwd(), 'bin', name, version)
  const binaryFileLocation = join(binDir, binaryInfo.name)
  if (!exists(binDir)) {
    createFolder(binDir, true)
    const url = getPlatformSpecificUrl(binaryInfo)

    console.log(`Downloading ${name} ${version} from ${url}...`)
    const response = await axios({
      url: url,
      method: 'GET',
      responseType: 'stream',
    })

    console.log(`Saving to ${binaryFileLocation}`)
    if (binaryInfo.type === 'tar.gz') {
      const extractorStream = response.data.pipe(createGunzip())
        .pipe(extract(binDir, {
          map: function (header) {
            const filename = basename(header.name)
            if (binaryInfo.files.includes(filename)) {
              // don't include folders when extracting files we want
              header.name = filename
            }
            return header
          },
          ignore: function (name) {
            return !binaryInfo.files.includes(basename(name))
          },
        }))
      return new Promise((resolve, reject) => {
        extractorStream.on('finish', () => {
          console.log('Done')
          resolve()
        })
        extractorStream.on('error', reject)
      })
    } else {
      const writer = createWriteStream(binaryFileLocation, { mode: 0o755 })
      response.data.pipe(writer)
      return new Promise((resolve, reject) => {
        writer.on('finish', () => {
          console.log('Done')
          resolve()
        })
        writer.on('error', reject)
      })
    }
  } else {
    console.log('Using cached binary at:', binaryFileLocation)
  }
}
