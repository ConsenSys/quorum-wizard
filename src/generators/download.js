import { basename } from 'path'
import axios from 'axios'
import { createGunzip } from 'zlib'
import { extract } from 'tar-fs'
import { createWriteStream } from 'fs'
import {
  createFolder,
  exists,
  wizardHomeDir,
} from '../utils/fileUtils'
import { info } from '../utils/log'
import { joinPath } from '../utils/pathUtils'

export async function getVersions(name) {
  const url = `https://api.bintray.com/packages/quorumengineering/${name}`
  const response = await axios({
    url,
    method: 'GET',
  })
  return response.data.versions
}

// eslint-disable-next-line import/prefer-default-export
export async function downloadIfMissing(name, version) {
  let binaryInfo
  switch (name) {
    case 'quorum':
      binaryInfo = createQuorumBinaryInfo(version)
      break
    case 'istanbul':
      binaryInfo = createIstanbulBinaryInfo(version)
      break
    case 'bootnode':
      binaryInfo = createBootnodeBinaryInfo(version)
      break
    default:
      binaryInfo = BINARIES[name][version]
  }
  const binDir = joinPath(wizardHomeDir(), 'bin', name, version)
  const binaryFileLocation = joinPath(binDir, binaryInfo.name)
  if (!exists(binaryFileLocation)) {
    createFolder(binDir, true)
    const url = getPlatformSpecificUrl(binaryInfo)

    info(`Downloading ${name} ${version} from ${url}...`)
    const response = await axios({
      url,
      method: 'GET',
      responseType: 'stream',
    })

    info(`Unpacking to ${binaryFileLocation}`)
    if (binaryInfo.type === 'tar.gz') {
      const extractorStream = response.data.pipe(createGunzip())
        .pipe(extract(binDir, {
          map: (header) => {
            const filename = basename(header.name)
            if (binaryInfo.files.includes(filename)) {
              // don't include folders when extracting files we want
              header.name = filename // eslint-disable-line no-param-reassign
            }
            return header
          },
          ignore: (pathName) => !binaryInfo.files.includes(basename(pathName)),
        }))
      return new Promise((resolve, reject) => {
        extractorStream.on('finish', () => {
          info(`Saved to ${binaryFileLocation}`)
          resolve()
        })
        extractorStream.on('error', reject)
      })
    }
    const writer = createWriteStream(binaryFileLocation, { mode: 0o755 })
    response.data.pipe(writer)
    return new Promise((resolve, reject) => {
      writer.on('finish', () => {
        info(`Saved to ${binaryFileLocation}`)
        resolve()
      })
      writer.on('error', reject)
    })
  }
  info(`Using cached ${name} at: ${binaryFileLocation}`)
  return binaryFileLocation
}

export function getPlatformSpecificUrl({ url }) {
  if (typeof url === 'string') {
    return url
  }
  const platformUrl = url[process.platform]
  if (platformUrl === undefined) {
    throw new Error(
      `Sorry, your platform (${process.platform}) is not supported.`,
    )
  }
  return platformUrl
}

export function createQuorumBinaryInfo(version) {
  return {
    name: 'geth',
    description: `Quorum ${version}`,
    url: {
      darwin: `https://bintray.com/quorumengineering/quorum/download_file?file_path=${version}/geth_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/quorumengineering/quorum/download_file?file_path=${version}/geth_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'geth',
    ],
  }
}

export function createIstanbulBinaryInfo(version) {
  return {
    name: 'istanbul',
    url: {
      darwin: `https://bintray.com/api/ui/download/quorumengineering/istanbul-tools/istanbul-tools_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/api/ui/download/quorumengineering/istanbul-tools/istanbul-tools_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'istanbul',
    ],
  }
}

export function createBootnodeBinaryInfo(version) {
  return {
    name: 'bootnode',
    url: {
      darwin: `https://bintray.com/api/ui/download/quorumengineering/geth-bootnode/bootnode_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/api/ui/download/quorumengineering/geth-bootnode/bootnode_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'bootnode',
    ],
  }
}

export const BINARIES = {
  tessera: {
    '0.10.4': {
      name: 'tessera-app.jar',
      description: 'Tessera 0.10.4',
      url: 'https://oss.sonatype.org/service/local/repositories/releases/content/com/jpmorgan/quorum/tessera-app/0.10.4/tessera-app-0.10.4-app.jar',
      type: 'jar',
    },
    '0.10.2': {
      name: 'tessera-app.jar',
      description: 'Tessera 0.10.2',
      url: 'https://oss.sonatype.org/service/local/repositories/releases/content/com/jpmorgan/quorum/tessera-app/0.10.2/tessera-app-0.10.2-app.jar',
      type: 'jar8',
    },
  },

  cakeshop: {
    '0.11.0': {
      name: 'cakeshop.war',
      description: 'Cakeshop 0.11.0',
      url: 'https://github.com/jpmorganchase/cakeshop/releases/download/v0.11.0/cakeshop-0.11.0.war',
      type: 'jar',
    },
    '0.11.0-J8': {
      name: 'cakeshop.war',
      description: 'Cakeshop 0.11.0-J8',
      url: 'https://github.com/jpmorganchase/cakeshop/releases/download/v0.11.0/cakeshop-0.11.0-J8.war',
      type: 'jar8',
    },
  },
}
