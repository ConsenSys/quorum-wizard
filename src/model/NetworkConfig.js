import { isJava11Plus } from '../utils/execUtils'
import {
  LATEST_CAKESHOP,
  LATEST_CAKESHOP_J8,
  LATEST_QUORUM,
  LATEST_TESSERA,
  LATEST_TESSERA_J8,
} from '../generators/download'
import {
  getDockerSubnet,
  cidrhost,
} from '../utils/subnetUtils'
import { getOutputPath } from '../utils/fileUtils'

export function createConfigFromAnswers(answers) {
  const {
    name,
    networkPath = getOutputPath(),
    numberNodes = 3,
    consensus = 'raft',
    quorumVersion = LATEST_QUORUM,
    transactionManager = isJava11Plus() ? LATEST_TESSERA : LATEST_TESSERA_J8,
    deployment = 'bash',
    cakeshop = isJava11Plus() ? LATEST_CAKESHOP : LATEST_CAKESHOP_J8,
    generateKeys = false,
    networkId = '10',
    genesisLocation = 'none',
    customizePorts = false,
    nodes = [],
    cakeshopPort = '8999',
    remoteDebug = false,
    containerPorts = undefined,
  } = answers
  const networkFolder = name
    || defaultNetworkName(numberNodes, consensus, transactionManager, deployment)
  const dockerSubnet = (isDocker(deployment) && containerPorts !== undefined) ? containerPorts.dockerSubnet : ''
  return {
    network: {
      name: networkFolder,
      networkPath,
      verbosity: 5,
      consensus,
      quorumVersion,
      transactionManager,
      permissioned: true,
      genesisFile: genesisLocation,
      generateKeys,
      configDir: `network/${networkFolder}/resources`,
      deployment,
      cakeshop,
      networkId,
      customizePorts,
      cakeshopPort,
      remoteDebug,
    },
    nodes: (customizePorts && nodes.length > 0) ? nodes : generateNodeConfigs(
      numberNodes,
      transactionManager,
      deployment,
      cakeshop,
      dockerSubnet,
    ),
    containerPorts,
  }
}

export function defaultNetworkName(numberNodes, consensus, transactionManager, deployment) {
  const transactionManagerName = !isTessera(transactionManager)
    ? ''
    : 'tessera-'
  return `${numberNodes}-nodes-${consensus}-${transactionManagerName}${deployment}`
}

export function generateNodeConfigs(
  numberNodes,
  transactionManager,
  deployment,
  cakeshop,
  dockerSubnet,
) {
  const devP2pPort = 21000
  const rpcPort = 22000
  const wsPort = 23000
  const graphQlPort = 24000
  const raftPort = 50401
  const thirdPartyPort = 9081
  const p2pPort = 9001
  const nodes = []

  for (let i = 0; i < parseInt(numberNodes, 10); i += 1) {
    const node = {
      quorum: {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 10) : '127.0.0.1',
        devP2pPort: devP2pPort + i,
        rpcPort: rpcPort + i,
        wsPort: wsPort + i,
        raftPort: raftPort + i,
        graphQlPort: graphQlPort + i,
      },
    }
    if (isTessera(transactionManager)) {
      node.tm = {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 100) : '127.0.0.1',
        thirdPartyPort: thirdPartyPort + i,
        p2pPort: p2pPort + i,
      }
    }
    nodes.push(node)
  }
  return nodes
}

export function getContainerPorts(deployment) {
  const dockerSubnet = isDocker(deployment) ? getDockerSubnet() : ''
  return {
    dockerSubnet,
    quorum: {
      rpcPort: 8545,
      p2pPort: 21000,
      raftPort: 50400,
      wsPort: 8645,
      graphQlPort: 8547,
    },
    tm: {
      p2pPort: 9000,
      thirdPartyPort: 9080,
    },
  }
}

export function isTessera(tessera) {
  return tessera !== 'none'
}

export function isDocker(deployment) {
  return deployment === 'docker-compose'
}

export function isBash(deployment) {
  return deployment === 'bash'
}

export function isKubernetes(deployment) {
  return deployment === 'kubernetes'
}

export function isIstanbul(consensus) {
  return consensus === 'istanbul'
}

export function isRaft(consensus) {
  return consensus === 'raft'
}

export function isCakeshop(cakeshop) {
  return cakeshop !== 'none'
}
