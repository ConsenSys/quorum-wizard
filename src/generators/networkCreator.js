import sanitize from 'sanitize-filename'
import {
  copyFile,
  createFolder,
  cwd,
  libRootDir,
  readFileToString,
  removeFolder,
  writeJsonFile,
  copyDirectory,
  writeFile,
  readDir,
  readJsonFile,
} from '../utils/fileUtils'
import { generateKeys } from './keyGen'
import { generateConsensusConfig } from '../model/ConsensusConfig'
import { createConfig } from '../model/TesseraConfig'
import {
  buildKubernetesResource,
  LATEST_QUBERNETES,
} from '../model/ResourceConfig'
import {
  isRaft,
  isTessera,
  isDocker,
  isKubernetes,
  isBash,
} from '../model/NetworkConfig'
import { joinPath } from '../utils/pathUtils'
import { executeSync } from '../utils/execUtils'
import { info } from '../utils/log'
import { buildDockerIp } from '../utils/subnetUtils'

export function createNetwork(config) {
  info('Building network directory...')
  const networkPath = getFullNetworkPath(config)
  removeFolder(networkPath)
  createFolder(networkPath, true)
  const configPath = getConfigPath()
  createFolder(configPath, true)
  writeJsonFile(configPath, `${config.network.name}-config.json`, config)
}

export function generateResourcesRemote(config) {
  info('Pulling docker container and generating network resources...')
  const configDir = joinPath(cwd(), config.network.configDir)
  const networkPath = getFullNetworkPath(config)
  const remoteOutputDir = joinPath(networkPath, 'out', 'config')

  const file = buildKubernetesResource(config)
  writeFile(joinPath(networkPath, 'qubernetes.yaml'), file, false)

  const initScript = isKubernetes(config.network.deployment) ? 'qube-init' : 'quorum-init'
  const copy7nodes = !config.network.generateKeys ? 'cp -r /qubernetes/7nodes /qubernetes/out/config; ' : ''
  let dockerCommand = `cd ${networkPath}
  ## make sure docker is installed
  docker ps > /dev/null
  EXIT_CODE=$?

  if [ $EXIT_CODE -ne 0 ];
  then
    exit $EXIT_CODE
  fi

  docker pull quorumengineering/qubernetes:${LATEST_QUBERNETES}


  docker run --rm -v ${networkPath}/qubernetes.yaml:/qubernetes/qubernetes.yaml -v ${networkPath}/out:/qubernetes/out quorumengineering/qubernetes:${LATEST_QUBERNETES} /bin/bash -c "${copy7nodes}./${initScript} --action=update qubernetes.yaml"
  `

  if (isDocker(config.network.deployment)) {
    dockerCommand += `
    sed -i'.bak' 's/%QUORUM-NODE\\([0-9]\\)_SERVICE_HOST%/${buildDockerIp(config.containerPorts.dockerSubnet, '1')}\\1/g' ${networkPath}/out/config/permissioned-nodes.json`
  }

  try {
    executeSync(dockerCommand)
  } catch (e) {
    throw new Error('Remote generation failed')
  }
  if (isDocker(config.network.deployment)) {
    copyDirectory(remoteOutputDir, configDir)
  }
}

export async function generateResourcesLocally(config) {
  info('Generating network resources locally...')
  const configDir = joinPath(cwd(), config.network.configDir)
  createFolder(configDir, true)

  if (config.network.generateKeys) {
    await generateKeys(config, configDir)
  } else {
    copyDirectory(joinPath(libRootDir(), '7nodes'), configDir)
  }

  generateConsensusConfig(
    configDir,
    config.network.consensus,
    config.nodes,
    config.network.networkId,
  )

  const staticNodes = createStaticNodes(config.nodes, config.network.consensus, configDir)
  writeJsonFile(configDir, 'permissioned-nodes.json', staticNodes)
}

export function createQdataDirectory(config) {
  // https://nodejs.org/en/knowledge/file-system/security/introduction/
  info('Building qdata directory...')
  const networkPath = getFullNetworkPath(config)
  const qdata = joinPath(networkPath, 'qdata')
  const logs = joinPath(qdata, 'logs')
  createFolder(logs, true)

  const configPath = joinPath(cwd(), config.network.configDir)

  const peerList = createPeerList(config.nodes, config.network.transactionManager)

  config.nodes.forEach((node, i) => {
    const nodeNumber = i + 1
    const keySource = joinPath(configPath, `key${nodeNumber}`)
    const quorumDir = joinPath(qdata, `dd${nodeNumber}`)
    const gethDir = joinPath(quorumDir, 'geth')
    const keyDir = joinPath(quorumDir, 'keystore')
    const tmDir = joinPath(qdata, `c${nodeNumber}`)
    const passwordDestination = joinPath(keyDir, 'password.txt')
    const genesisDestination = joinPath(quorumDir, 'genesis.json')
    createFolder(quorumDir)
    createFolder(gethDir)
    createFolder(keyDir)

    copyFile(joinPath(configPath, 'permissioned-nodes.json'), joinPath(quorumDir, 'permissioned-nodes.json'))
    copyFile(joinPath(configPath, 'permissioned-nodes.json'), joinPath(quorumDir, 'static-nodes.json'))
    copyFile(joinPath(keySource, 'acctkeyfile.json'), joinPath(keyDir, 'key'))
    copyFile(joinPath(keySource, 'nodekey'), joinPath(gethDir, 'nodekey'))
    copyFile(joinPath(keySource, 'password.txt'), passwordDestination)
    copyFile(joinPath(configPath, 'genesis.json'), genesisDestination)
    if (isTessera(config.network.transactionManager)) {
      createFolder(tmDir)
      copyFile(joinPath(keySource, 'tm.key'), joinPath(tmDir, 'tm.key'))
      copyFile(joinPath(keySource, 'tm.pub'), joinPath(tmDir, 'tm.pub'))

      if (isBash(config.network.deployment)) {
        const tesseraConfig = createConfig(
          tmDir,
          nodeNumber,
          node.tm.ip,
          node.tm.thirdPartyPort,
          node.tm.p2pPort,
          peerList,
        )
        writeJsonFile(tmDir, `tessera-config-09-${nodeNumber}.json`, tesseraConfig)
      } else {
        copyFile(joinPath(configPath, 'tessera-config-9.0.json'), joinPath(tmDir, 'tessera-config-09.json'))
      }
    }
  })
}

export function createStaticNodes(nodes, consensus, configDir) {
  return nodes.map((node, i) => {
    const nodeNumber = i + 1
    const generatedKeyFolder = `${configDir}/key${nodeNumber}`
    const enodeId = readFileToString(joinPath(generatedKeyFolder, 'enode'))

    let enodeAddress = `enode://${enodeId}@${node.quorum.ip}:${node.quorum.devP2pPort}?discport=0`
    if (isRaft(consensus)) {
      enodeAddress += `&raftport=${node.quorum.raftPort}`
    }
    return enodeAddress
  })
}

function createPeerList(nodes, transactionManager) {
  if (!isTessera(transactionManager)) {
    return []
  }
  return nodes.map((node) => ({ url: `http://${node.tm.ip}:${node.tm.p2pPort}` }))
}

export function getFullNetworkPath(config) {
  const networkFolderName = sanitize(config.network.name)
  if (networkFolderName === '') {
    throw new Error('Network name was empty or contained invalid characters')
  }

  return joinPath(cwd(), 'network', networkFolderName)
}

function getConfigPath(...relativePaths) {
  return joinPath(cwd(), 'configs', ...relativePaths)
}

export function readConfigJson(...relativePaths) {
  return readJsonFile(getConfigPath(...relativePaths))
}
export function getAvailableConfigs() {
  const configDir = getConfigPath()
  const arr = readDir(configDir)
  return (arr && arr.length) ? arr : ['no configs available']
}
