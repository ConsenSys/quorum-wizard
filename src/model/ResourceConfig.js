import { isTessera } from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function buildKubernetesResource(config) {
  return [
    buildGeneralDetails(config),
    buildQuorumDetails(config),
    buildTesseraDetails(config),
    buildGethDetails(config),
  ].join('')
}

function buildGeneralDetails(config) {
  const nodeNum = config.nodes.length
  return `#namespace:
#  name: quorum-test
sep_deployment_files: true
generate_keys: ${config.network.generateKeys}
nodes:
  number: ${nodeNum}
service:
  type: NodePort`
}

function buildQuorumDetails(config) {
  return `
quorum:
  consensus: ${config.network.consensus}
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: ${config.nodes[0].quorum.raftPort}
    Quorum_Version: ${config.network.quorumVersion}
  storage:
    Type: PVC
    Capacity: 200Mi`
}

function buildGethDetails(config) {
  return `
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 30303
  network:
    id: ${config.network.networkId}
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\"*\\"`
}

function buildTesseraDetails(config) {
  const hasTessera = isTessera(config.network.transactionManager)
  return hasTessera
    ? `
  tm:
    Name: tessera
    Tm_Version: ${config.network.transactionManager}
    Port: ${config.nodes[0].tm.thirdPartyPort}
    Tessera_Config_Dir: out/config`
    : ''
}
