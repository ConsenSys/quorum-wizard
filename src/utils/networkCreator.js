import { join, normalize } from 'path'
import { execute } from './execUtils'
import sanitize from 'sanitize-filename'
import {
  copyFile,
  createFolder,
  readFileToString,
  removeFolder,
  writeFile,
  writeJsonFile,
} from './fileUtils'
import { generateKeys } from './keyGen'
import { generateConsensusConfig } from '../model/ConsensusConfig'
import { createConfig } from '../model/TesseraConfig'
import { isBash } from '../model/NetworkConfig'
import { createGethStartCommand, createTesseraStartCommand, waitForTesseraNodesCommand } from './bashHelper'

export function createDirectory (config) {
  // https://nodejs.org/en/knowledge/file-system/security/introduction/

  const networkFolderName = sanitize(config.network.name)
  if (networkFolderName === '') {
    throw new Error('Network name was empty or contained invalid characters')
  }

  const networkPath = join(process.cwd(), 'network', networkFolderName)
  removeFolder(networkPath)

  const qdata = join(networkPath, 'qdata')
  const logs = join(qdata, 'logs')
  createFolder(logs, true)
  writeJsonFile(networkPath, 'config.json', config)

  const configPath = join(process.cwd(), config.network.configDir)
  if(config.network.generateKeys) {
      generateKeys(config, configPath)
      generateConsensusConfig(configPath, config.network.consensus, config.nodes)
  }

  const staticNodes = createStaticNodes(config.nodes, config.network.consensus, config.network.configDir)
  const peerList = createPeerList(config.nodes, config.network.transactionManager)
  const initCommands = []
  const startCommands = []
  const tmStartCommands = []

  config.nodes.forEach((node, i) => {
    const nodeNumber = i + 1
    const keyFolder = join(configPath, `key${nodeNumber}`)
    const quorumDir = join(qdata, `dd${nodeNumber}`)
    const gethDir = join(quorumDir, 'geth')
    const keyDir = join(quorumDir, 'keystore')
    const tmDir = join(qdata, `c${nodeNumber}`)
    const passwordDestination = join(keyDir, 'password.txt')
    const genesisDestination = join(quorumDir, `${config.network.consensus}-genesis.json`)
    createFolder(quorumDir)
    createFolder(gethDir)
    createFolder(keyDir)
    createFolder(tmDir)

    writeJsonFile(quorumDir, 'permissioned-nodes.json', staticNodes)
    writeJsonFile(quorumDir, 'static-nodes.json', staticNodes)
    copyFile(normalize(config.network.genesisFile), genesisDestination)
    copyFile(join(keyFolder, 'key'), join(keyDir, 'key'))
    copyFile(join(keyFolder, 'nodekey'), join(gethDir, 'nodekey'))
    copyFile(join(keyFolder, 'password.txt'), passwordDestination)
    copyFile(join(configPath, `${config.network.consensus}-genesis.json`), genesisDestination)
    if(isTessera(config)) {
      copyFile(join(keyFolder, 'tm.key'), join(tmDir, 'tm.key'))
      copyFile(join(keyFolder, 'tm.pub'), join(tmDir, 'tm.pub'))
      let tesseraConfig = createConfig(tmDir, nodeNumber,
        node.tm.thirdPartyPort, node.tm.p2pPort, peerList)
      writeJsonFile(tmDir, `tessera-config-09-${nodeNumber}.json`,
        tesseraConfig)
    }

    if (config.network.deployment === 'bash') {
      const initCommand = `cd ${networkPath} && geth --datadir ${quorumDir} init ${genesisDestination}`
      initCommands.push(initCommand)

      let tmIpcLocation = isTessera(config) ? join(tmDir, 'tm.ipc') : 'ignore'
      const startCommand = createGethStartCommand(config, node,
        passwordDestination, nodeNumber, tmIpcLocation)
        startCommands.push(startCommand)

      if (isTessera(config)) {
        const tmStartCommand = createTesseraStartCommand(config, node, nodeNumber,
          tmDir, logs)
        tmStartCommands.push(tmStartCommand)
      }
    }
  })

  var obj = {
    tesseraStart:  tmStartCommands.join('\n'),
    gethStart: startCommands.join('\n'),
    initStart: initCommands,
    netPath: networkPath,
  }
  return obj;
}

export function createStaticNodes (nodes, consensus, configDir) {
  return nodes.map((node, i) => {
    const nodeNumber = i + 1
    const generatedKeyFolder = `${configDir}/key${nodeNumber}`
    const enodeId = readFileToString(join(generatedKeyFolder, 'enode'))

    let enodeAddress = `enode://${enodeId}@${node.quorum.ip}:${node.quorum.devP2pPort}?discport=0`
    if (consensus === 'raft') {
      enodeAddress += `&raftport=${node.quorum.raftPort}`
    }
    return enodeAddress
  })
}

export function isTessera (config) {
  return config.network.transactionManager === 'tessera'
}

function createPeerList (nodes, transactionManager) {
  if (transactionManager !== 'tessera') {
    return []
  }
  return nodes.map((node) => ({
    url: `http://${node.tm.ip}:${node.tm.p2pPort}`
  }))
}
