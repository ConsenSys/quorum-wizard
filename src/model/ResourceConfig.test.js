import { getFullNetworkPath } from '../generators/networkCreator'
import { createConfigFromAnswers } from './NetworkConfig'
import {
  TEST_CWD,
} from '../utils/testHelper'
import { buildKubernetesResource } from './ResourceConfig'
import { LATEST_QUORUM, LATEST_TESSERA } from '../generators/download'

jest.mock('../utils/fileUtils')
jest.mock('../generators/networkCreator')
getFullNetworkPath.mockReturnValue(`${TEST_CWD}/test-network`)

const containerPortInfo = {
  quorum: {
    rpcPort: 8545,
    p2pPort: 21000,
    raftPort: 50400,
    wsPort: 8645,
  },
  tm: {
    p2pPort: 9000,
    thirdPartyPort: 9080,
  },
}
const baseNetwork = {
  numberNodes: '5',
  consensus: 'raft',
  quorumVersion: LATEST_QUORUM,
  transactionManager: LATEST_TESSERA,
  cakeshop: 'none',
  deployment: 'kubernetes',
  containerPorts: {
    dockerSubnet: '',
    ...containerPortInfo,
  },
}

test('creates 5nodes raft kubernetes tessera', () => {
  const config = createConfigFromAnswers(baseNetwork)
  const kubernetes = buildKubernetesResource(config)
  expect(kubernetes).toMatchSnapshot()
})

test('creates 7nodes istanbul kubernetes generate keys', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    numberNodes: '7',
    consensus: 'istanbul',
    generateKeys: true,
  })
  const kubernetes = buildKubernetesResource(config)
  expect(kubernetes).toMatchSnapshot()
})

test('creates 7nodes istanbul docker generate keys', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    numberNodes: '7',
    deployment: 'docker-compose',
    consensus: 'istanbul',
    generateKeys: true,
    containerPorts: {
      dockerSubnet: '172.16.239.0/24',
      ...containerPortInfo,
    },
  })
  const kubernetes = buildKubernetesResource(config)
  expect(kubernetes).toMatchSnapshot()
})

test('creates 7nodes istanbul docker generate keys no tessera', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    numberNodes: '7',
    deployment: 'docker-compose',
    consensus: 'istanbul',
    transactionManger: 'none',
    generateKeys: true,
    containerPorts: {
      dockerSubnet: '172.16.239.0/24',
      ...containerPortInfo,
    },
  })
  const kubernetes = buildKubernetesResource(config)
  expect(kubernetes).toMatchSnapshot()
})
