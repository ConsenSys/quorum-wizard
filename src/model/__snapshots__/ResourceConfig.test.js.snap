// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates 5nodes raft kubernetes tessera 1`] = `
"#namespace:
#  name: quorum-test
sep_deployment_files: true
nodes:
  number: 5
service:
  type: NodePort
quorum:
  consensus: raft
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: 7nodes
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: 50400
    Quorum_Version: 2.6.0
  storage:
    Type: PVC
    Capacity: 200Mi
  tm:
    Name: tessera
    Tm_Version: 0.10.5
    Port: 9000
    3Party_Port: 9080
    Tessera_Config_Dir: out/config
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 21000
  network:
    id: 10
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\\\\\"*\\\\\\""
`;

exports[`creates 7nodes istanbul docker generate keys 1`] = `
"#namespace:
#  name: quorum-test
sep_deployment_files: true
nodes:
  number: 7
service:
  type: NodePort
quorum:
  consensus: istanbul
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: 50400
    Quorum_Version: 2.6.0
  storage:
    Type: PVC
    Capacity: 200Mi
  tm:
    Name: tessera
    Tm_Version: 0.10.5
    Port: 9000
    3Party_Port: 9080
    Tessera_Config_Dir: out/config
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 21000
  network:
    id: 10
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\\\\\"*\\\\\\""
`;

exports[`creates 7nodes istanbul docker generate keys no tessera 1`] = `
"#namespace:
#  name: quorum-test
sep_deployment_files: true
nodes:
  number: 7
service:
  type: NodePort
quorum:
  consensus: istanbul
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: 50400
    Quorum_Version: 2.6.0
  storage:
    Type: PVC
    Capacity: 200Mi
  tm:
    Name: tessera
    Tm_Version: 0.10.5
    Port: 9000
    3Party_Port: 9080
    Tessera_Config_Dir: out/config
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 21000
  network:
    id: 10
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\\\\\"*\\\\\\""
`;

exports[`creates 7nodes istanbul kubernetes generate keys 1`] = `
"#namespace:
#  name: quorum-test
sep_deployment_files: true
nodes:
  number: 7
service:
  type: NodePort
quorum:
  consensus: istanbul
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: 50400
    Quorum_Version: 2.6.0
  storage:
    Type: PVC
    Capacity: 200Mi
  tm:
    Name: tessera
    Tm_Version: 0.10.5
    Port: 9000
    3Party_Port: 9080
    Tessera_Config_Dir: out/config
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 21000
  network:
    id: 10
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\\\\\"*\\\\\\""
`;
